src/cargo/ops/cargo_install.rs

```diff
@@ -442,9 +442,11 @@ impl<'gctx> InstallablePackage<'gctx> {
             let src = staging_dir.path().join(bin);
             let dst = dst.join(bin);
             self.gctx.shell().status("Installing", dst.display())?;
-            fs::rename(&src, &dst).with_context(|| {
-                format!("failed to move `{}` to `{}`", src.display(), dst.display())
-            })?;
+            if !dry_run {
+                fs::rename(&src, &dst).with_context(|| {
+                    format!("failed to move `{}` to `{}`", src.display(), dst.display())
+                })?;
+            }
             installed.bins.push(dst);
             successful_bins.insert(bin.to_string());
         }
```

Comment:
Can't highlight it but we also need to have dry-run affect
- `replace_to`s rename (should also get a test)
- orphan removal (would be nice to have a test)

Reply:
Thanks for calling this out. Coming from <https://github.com/rust-lang/cargo/issues/11123#issuecomment-1910469992> we need to answer these user questions. I think the first item is the only one left.


- [ ] How: Is it an update (replace), a fresh install, or no-op?
- [x] What: which binary from which source is going to be installed
- [x] Where: to where the binary will be installed